### PR TITLE
Sample Difficulty Curriculum: oversample hard training cases

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The training set contains easy and hard samples. Easy samples (low AoA, single foil, mid-range Re) dominate SGD updates but contribute little learning once fit. **Oversampling the hardest training samples** (highest recent loss) redirects gradient budget toward cases where the model needs most improvement.

Different from: Re-stratified sampling (targets Re distribution), hard-node mining (targets hard NODES), PCGrad extreme (gradient conflicts).

### Changes to `cfd_tandemfoil/train.py`

1. **Add flags:**
   ```python
   parser.add_argument('--sample_curriculum', action='store_true')
   parser.add_argument('--curriculum_beta', type=float, default=2.0, help='Hard sample oversampling factor')
   parser.add_argument('--curriculum_warmup', type=int, default=10, help='Epochs before curriculum activates')
   ```

2. **Maintain a per-sample loss tracker:**
   ```python
   if args.sample_curriculum:
       n_train = len(train_dataset)
       sample_losses = torch.zeros(n_train)
       sample_counts = torch.zeros(n_train)
   ```

3. **Update after each batch** (need per-sample loss and sample indices):
   ```python
   if args.sample_curriculum:
       batch_indices = batch['idx']  # ensure dataloader returns indices
       per_sample_loss = loss_per_sample.detach().cpu()
       alpha = 0.3
       for i, idx in enumerate(batch_indices):
           if sample_counts[idx] == 0:
               sample_losses[idx] = per_sample_loss[i]
           else:
               sample_losses[idx] = (1-alpha)*sample_losses[idx] + alpha*per_sample_loss[i]
           sample_counts[idx] += 1
   ```

4. **Update sampler at each epoch start (after warmup):**
   ```python
   if args.sample_curriculum and epoch >= args.curriculum_warmup:
       valid = sample_counts > 0
       weights = torch.ones(n_train)
       if valid.any():
           ln = sample_losses[valid]
           ln = (ln - ln.min()) / (ln.max() - ln.min() + 1e-8)
           weights[valid] = 1.0 + (args.curriculum_beta - 1.0) * ln
       sampler = WeightedRandomSampler(weights, num_samples=n_train, replacement=True)
       train_loader = DataLoader(train_dataset, batch_size=batch_size, sampler=sampler, ...)
   ```

5. **Important:** Multiply curriculum weights with existing re_stratified_sampling weights (complementary). Ensure dataloader returns sample indices. Log `train/curriculum_weight_std` and `train/curriculum_weight_max` to W&B.

6. **Run 2 seeds:**
   ```bash
   CUDA_VISIBLE_DEVICES=0 python train.py \
     --agent nezuko --wandb_name "nezuko/curriculum-b2.0-s42" --wandb_group sample-difficulty-curriculum \
     --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
     --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
     --pressure_first --pressure_deep --residual_prediction --surface_refine \
     --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
     --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
     --sample_curriculum --curriculum_beta 2.0 --curriculum_warmup 10

   CUDA_VISIBLE_DEVICES=1 python train.py \
     --agent nezuko --wandb_name "nezuko/curriculum-b2.0-s73" --wandb_group sample-difficulty-curriculum \
     --seed 73 [same flags as above]
   ```

## Baseline
| Metric | 2-seed avg | Target |
|--------|-----------|--------|
| **p_in** | **11.709** | < 11.71 |
| **p_oodc** | **7.544** | < 7.54 |
| **p_tan** | **27.402** | < 27.40 |
| p_re | 6.481 | < 6.48 |

W&B baseline: h6fqcry4 (s42), cuhoscp9 (s73)
Reproduce: `cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT nezuko:

## Results

### Sample Difficulty Curriculum: oversample hard training cases (beta=2.0, warmup=10)

**Verdict: Neutral — no meaningful improvement over baseline.** Metrics are within noise of baseline with slight regressions on p_in and p_re, slight improvements on p_oodc and p_tan.

### Metrics (2-seed avg)

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline | Δ |
|--------|---------|---------|------------|----------|---|
| p_in   | 12.3    | 12.1    | **12.20**  | 11.709   | **+4.2%** ❌ |
| p_oodc | 7.7     | 7.3     | **7.50**   | 7.544    | **-0.6%** ✅ |
| p_tan  | 27.4    | 27.2    | **27.30**  | 27.402   | **-0.4%** ✅ |
| p_re   | 6.7     | 6.3     | **6.50**   | 6.481    | **+0.3%** ❌ |

### Experiment details

**Commands:**
```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent nezuko --wandb_name "nezuko/curriculum-b2.0-s42" --wandb_group sample-difficulty-curriculum \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
  --sample_curriculum --curriculum_beta 2.0 --curriculum_warmup 10
# Seed 73: same with --seed 73 and CUDA_VISIBLE_DEVICES=1
```

- **W&B runs:** qyao6jov (seed 42), wfuyfg03 (seed 73)
- **Epochs:** ~160 (hit timeout — comparable to baseline ~159)
- **Peak VRAM:** ~76 GB (s42), ~87 GB (s73) — similar to baseline

### Implementation

1. Created `_IndexedDataset` wrapper class to return sample indices from the dataloader (read-only `pad_collate` can't be modified)
2. Custom `_collate_with_idx` wraps `pad_collate` to extract indices
3. Per-sample loss tracker using EMA (alpha=0.3) on `surf_per_sample` (surface pressure MAE per sample)
4. At each epoch start (after warmup), curriculum weights computed: `1.0 + (beta-1) * normalized_loss`, multiplied with base domain+Re-stratified weights
5. `WeightedRandomSampler.weights` updated in-place each epoch
6. Logged `train/curriculum_weight_std` and `train/curriculum_weight_max` to W&B

### What happened — why it didn't help

1. **The baseline already handles hard samples well.** The existing hard-node mining (epoch 30+), PCGrad 3-way gradient surgery, tandem adaptive boost, and Re-stratified sampling already target the same problem from multiple angles. Adding sample-level curriculum on top provides minimal additional signal.

2. **Curriculum weight differentiation was modest.** After warmup, curriculum weight std was ~0.04 with max=2.0. This means most samples got weights close to 1.0 — the loss landscape across samples is relatively uniform after the existing stratification. Only a small fraction of samples were meaningfully upweighted.

3. **p_in regression (+4.2%):** The curriculum may slightly hurt in-distribution performance by shifting gradient budget toward harder OOD samples (which already get attention from PCGrad and tandem boost). This is a known trade-off in curriculum learning.

### Suggested follow-ups

- **Higher beta (e.g. 5.0-10.0)** could amplify the effect, but risks destabilizing training by over-focusing on outliers
- **Per-field curriculum** (tracking pressure loss separately from velocity) might be more targeted
- The marginal improvement on p_oodc and p_tan suggests the approach has *some* value for OOD generalization — a variant that only upweights OOD-like samples (high AoA, extreme gap/stagger) might work better
- Given the existing multi-layered difficulty-targeting (hard-node mining + PCGrad + tandem boost + Re-stratification), diminishing returns from yet another layer are expected

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2347 because ADVISOR: Closing — Sample difficulty curriculum is approximately neutral (p_in +4.2%, p_oodc -0.6%, p_tan -0.4%, p_re +0.3%). The baseline already handles difficulty through hard-node mining, PCGrad 3-way, tandem boost, and Re-stratified sampling. Curriculum weight differentiation was modest (std ~0.04) — the loss landscape across samples is already relatively uniform after existing stratification. Diminishing returns from yet another difficulty-targeting layer..
